controller/api/destination/server.go

```diff
@@ -38,185 +41,141 @@ type server struct {
-
-	proxyNS := ""
-	parts := strings.Split(dest.GetContextToken(), ":")
-	// ns:<namespace>
-	if len(parts) == 2 && parts[0] == "ns" {
-		proxyNS = parts[1]
-	} else {
-		// TODO remove this after the 2.3 release...
-		parts = strings.Split(dest.GetContextToken(), ".")
-		// <deployment>.deployment.<namespace>.linkerd-managed.linkerd.svc.cluster.local
-		if len(parts) >= 3 {
-			log.Debug("Serving profile request for legacy proxy")
-			proxyNS = parts[2]
-		}
-	}
-	if proxyNS != "" {
-		log.Debugf("Looking up profile given context: ns:%s", proxyNS)
+	select {
+	case <-s.shutdown:
+	case <-stream.Context().Done():
+		log.Debugf("Get %s cancelled", dest.GetPath())
 	}
 
-	err = s.resolver.streamProfiles(host, proxyNS, listener)
-	if err != nil {
-		s.log.Errorf("Error streaming profile for %s: %v", dest.Path, err)
-	}
-	return err
+	return nil
 }
 
-func (s *server) Endpoints(ctx context.Context, params *discoveryPb.EndpointsParams) (*discoveryPb.EndpointsResponse, error) {
-	s.log.Debugf("Endpoints(%+v)", params)
-
-	servicePorts := s.resolver.getState()
-
-	rsp := discoveryPb.EndpointsResponse{
-		ServicePorts: make(map[string]*discoveryPb.ServicePort),
-	}
-
-	for serviceID, portMap := range servicePorts {
-		discoverySP := discoveryPb.ServicePort{
-			PortEndpoints: make(map[uint32]*discoveryPb.PodAddresses),
-		}
-		for port, sp := range portMap {
-			podAddrs := discoveryPb.PodAddresses{
-				PodAddresses: []*discoveryPb.PodAddress{},
-			}
-
-			for _, ua := range sp.addresses {
-				ownerKind, ownerName := s.k8sAPI.GetOwnerKindAndName(ua.pod)
-				pod := util.K8sPodToPublicPod(*ua.pod, ownerKind, ownerName)
-
-				podAddrs.PodAddresses = append(
-					podAddrs.PodAddresses,
-					&discoveryPb.PodAddress{
-						Addr: addr.NetToPublic(ua.address),
-						Pod:  &pod,
-					},
-				)
-			}
-
-			discoverySP.PortEndpoints[port] = &podAddrs
-		}
-
-		s.log.Debugf("ServicePorts[%s]: %+v", serviceID, discoverySP)
-		rsp.ServicePorts[serviceID.String()] = &discoverySP
+func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetProfileServer) error {
+	log := s.log
+	client, _ := peer.FromContext(stream.Context())
+	if client != nil {
+		log = log.WithField("remote", client.Addr)
 	}
+	log.Debugf("GetProfile(%+v)", dest)
 
-	return &rsp, nil
-}
+	translator := newProfileTranslator(stream, log)
 
-func (s *server) streamResolution(host string, port int, stream pb.Destination_GetServer) error {
-	listener := newEndpointListener(stream, s.k8sAPI.GetOwnerKindAndName, s.enableH2Upgrade, s.controllerNS, s.identityTrustDomain)
+	primary, secondary := newFallbackProfileListener(translator)
 
-	resolverCanResolve, err := s.resolver.canResolve(host, port)
-	if err != nil {
-		return fmt.Errorf("resolver [%+v] found error resolving host [%s] port [%d]: %v", s.resolver, host, port, err)
-	}
-	if !resolverCanResolve {
-		return fmt.Errorf("cannot find resolver for host [%s] port [%d]", host, port)
+	if dest.GetContextToken() != "" {
```

Comment:
It's hard to see it in the context of this diff, but if the context token exists, won't we end up calling subscribe twice, once with the context token, and once without it? I think that will result in sending duplicate updates to the proxy? Should it instead be:

```diff
diff --git a/controller/api/destination/server.go b/controller/api/destination/server.go
index 3fad31a9..13f36d3a 100644
--- a/controller/api/destination/server.go
+++ b/controller/api/destination/server.go
@@ -125,11 +125,11 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
 	if dest.GetContextToken() != "" {
 		s.profiles.Subscribe(dest.GetPath(), dest.GetContextToken(), primary)
 		defer s.profiles.Unsubscribe(dest.GetPath(), dest.GetContextToken(), primary)
+	} else {
+		s.profiles.Subscribe(dest.GetPath(), "", secondary)
+		defer s.profiles.Unsubscribe(dest.GetPath(), "", secondary)
 	}
 
-	s.profiles.Subscribe(dest.GetPath(), "", secondary)
-	defer s.profiles.Unsubscribe(dest.GetPath(), "", secondary)
-
 	select {
 	case <-s.shutdown:
 	case <-stream.Context().Done():
```

Reply:
It is intentional that we create two subscriptions here.  The fallbackProfileListener consumes updates from both subscriptions (via the `primary` and `secondary` listeners) and decides which updates to send to the proxy.  I'll add a comment here explaining this behavior.